src/java.base/unix/classes/sun/nio/fs/UnixPath.java

```diff
@@ -857,13 +860,83 @@ public Path toRealPath(LinkOption... options) throws IOException {
             result = result.resolve(element);
         }
 
-        // check file exists (without following links)
+        // check whether file exists (without following links)
         try {
             UnixFileAttributes.get(result, false);
         } catch (UnixException x) {
             x.rethrowAsIOException(result);
         }
-        return result;
+
+        // Return if the file system is not both case insensitive and retentive
+        if (!fs.isCaseInsensitiveAndRetentive())
+            return result;
+
+        UnixPath path = fs.rootDirectory();
+
+        SecurityManager sm = System.getSecurityManager();
+
+        // Traverse the result obtained above from the root downward, leaving
+        // any '..' elements intact, and replacing other elements with the
+        // entry in the same directory which has an equal key
+        for (int i = 0; i < result.getNameCount(); i++ ) {
+            UnixPath elt = result.getName(i);
+
+            // If the element is "..", append it directly and continue
+            if (elt.toString().equals("..")) {
+                path = path.resolve(elt);
+                continue;
+            }
+
+            // Derive full path to element and check readability
+            UnixPath eltPath = path.resolve(elt);
+            if (sm != null)
+                sm.checkRead(eltPath.getPathForPermissionCheck());
+
+            // Derive element key
+            UnixFileAttributes attrs = null;
+            try {
+                attrs = UnixFileAttributes.get(eltPath, false);
+            } catch (UnixException x) {
+                x.rethrowAsIOException(result);
+            }
+            final UnixFileKey eltKey = attrs.fileKey();
+
+            // Check readbility of path thus far
+            if (sm != null)
+                sm.checkRead(path.getPathForPermissionCheck());
+
+            // Filter entries whose UnixFileKey equals 'eltKey'
+            DirectoryStream.Filter<Path> filter = (p) -> {
+                UnixFileAttributes attributes = null;
+                try {
+                    attributes = UnixFileAttributes.get(toUnixPath(p), false);
+                } catch (UnixException x) {
+                    x.rethrowAsIOException(this);
```

Comment:
This is problematic as it means that toRealPath can fail trying to access a file that is not in the path. Also given that it is only looking for one file then it might be simpler to just handle in the loop rather than as a filter.

Reply:
Addressed in 49d18c6a39d9d70b7da91c5b4b3740ba3c890948.